dipy/workflows/align.py

```diff
@@ -45,16 +54,464 @@ def run(self, input_files, new_vox_size, order=1, mode='constant', cval=0,
             logging.info('Processing {0}'.format(inputfile))
             new_data, new_affine = reslice(data, affine, vox_sz, new_vox_size,
                                            order, mode=mode, cval=cval,
                                            num_processes=num_processes)
             save_nifti(outpfile, new_data, new_affine)
             logging.info('Resliced file save in {0}'.format(outpfile))
-        
\ No newline at end of file
+
+
+class ImageRegistrationFlow(Workflow):
+    """
+    The registration workflow is organized as a collection of different
+    functions. The user can intend to use only one type of registration
+    (such as center of mass or rigid body registration only).
+
+    Alternatively, a registration can be done in a progressive manner.
+    For example, using affine registration with progressive set to 'True'
+    will involve center of mass, translation, rigid body and full affine
+    registration. Whereas, when progressive is False the registration will
+    include only center of mass and affine registration. The progressive
+    registration will be slower but will improve the quality.
+
+    This can be controlled by using the progressive flag (True by default).
+    """
+
+    def center_of_mass(self, static, static_grid2world,
+                       moving, moving_grid2world):
+
+        """ Function for the center of mass based image
+        registration.
+
+        Parameters
+        ----------
+        static : array, shape (S, R, C) or (R, C)
+            the image to be used as reference during optimization.
+
+        static_grid2world : array, shape (dim+1, dim+1), optional
+            the voxel-to-space transformation associated with the static
+            image. The default is None, implying the transform is the
+            identity.
+
+        moving : array, shape (S', R', C') or (R', C')
+            the image to be used as "moving" during optimization. It is
+            necessary to pre-align the moving image to ensure its domain
+            lies inside the domain of the deformation fields. This is assumed
+            to be accomplished by "pre-aligning" the moving image towards the
+            static using an affine transformation given by the
+            'starting_affine' matrix
+
+        moving_grid2world : array, shape (dim+1, dim+1), optional
+            the voxel-to-space transformation associated with the moving
+            image. The default is None, implying the transform is the
+            identity.
+
+        """
+
+        img_registration = transform_centers_of_mass(static,
+                                                     static_grid2world,
+                                                     moving,
+                                                     moving_grid2world)
+
+        transformed = img_registration.transform(moving)
+        return transformed, img_registration.affine
+
+    def translate(self, static, static_grid2world, moving,
```

Comment:
There's clearly a pattern here. `translate`, `rigid` and `affine` all have the ~same block of code. Only the transform type changes. Maybe you could create a common function for these functions?

Reply:
@nilgoyette: You are correct, there is clearly a pattern here but the thing is that this pattern can get a bit more complicated depending on how the user wants to run the analysis. So, I simply separated each type of registration in its own function. These functions, in turn, are calling other dependent functions to generate some intermediate results (depending on the type of analysis the user wants to perform.) So, translate can call center of mass or rigid can call translate and center of mass or don't call them at all.

With a single function to abstract the low-level details, it will become a bit more redundant code because it will also call different functions (like it's being done now). 

But, I am thinking about this one more, if there is a way to further abstract away the code then I surely will do that. It will structure the code even more nicely.

Thanks.

Reply:
I have tried to reduce the code duplication by moving the common code from (translate, rigid and affine) functions in a separate function. Thanks for this review.